nuxeo-mongodb-directory/src/main/java/org/nuxeo/directory/MongoDBSession.java

```diff
@@ -0,0 +1,349 @@
+        for (Map.Entry<String, Serializable> entry : fieldMap.entrySet()) {
+            Object value = MongoDBSerializationHelper.valueToBson(entry.getValue());
+            if (value != null) {
+                if (fulltext.contains(entry.getKey())) {
+                    switch (substringMatchType) {
+                    case subany:
+                        bson.put(entry.getKey(), Pattern.compile(String.valueOf(value), Pattern.CASE_INSENSITIVE));
+                        break;
+                    case subinitial:
+                        bson.put(entry.getKey(),
+                                Pattern.compile('^' + String.valueOf(value), Pattern.CASE_INSENSITIVE));
+                        break;
+                    case subfinal:
+                        bson.put(entry.getKey(),
+                                Pattern.compile(String.valueOf(value) + '$', Pattern.CASE_INSENSITIVE));
+                        break;
+                    }
+                } else {
+                    bson.put(entry.getKey(), value);
+                }
+            }
+        }
+        return bson;
+    }
+
+    @Override
+    public void close() throws DirectoryException {
+        client.close();
+    }
+
+    @Override
+    public List<String> getProjection(Map<String, Serializable> filter, String columnName) throws DirectoryException {
+        return getProjection(filter, Collections.emptySet(), columnName);
+    }
+
+    @Override
+    public List<String> getProjection(Map<String, Serializable> filter, Set<String> fulltext, String columnName)
+            throws DirectoryException {
+        DocumentModelList docList = query(filter, fulltext);
+        List<String> result = new ArrayList<>();
+        for (DocumentModel docModel : docList) {
+            Object obj = docModel.getProperty(schemaName, columnName);
+            String propValue = String.valueOf(obj);
+            result.add(propValue);
+        }
+        return result;
+    }
+
+    @Override
+    public boolean authenticate(String username, String password) throws DirectoryException {
+        // As the MongoClient do not throw any connectivity-related exceptions,
+        // a timeout will occur if a command is launched with bad username/password
+        MongoClient mongoClient = new MongoClient(client.getAddress(),
+                Arrays.asList(MongoCredential.createCredential(username, dbName, password.toCharArray())),
+                MongoClientOptions.builder().serverSelectionTimeout(1000).build());
+        try {
+            mongoClient.getDatabase(dbName).runCommand(new Document("ping", 1));
+            return true;
+        } catch (MongoTimeoutException e) {
+            return false;
+        } finally {
+            mongoClient.close();
+        }
+    }
+
+    @Override
+    public boolean hasEntry(String id) {
+        return getCollection().count(
+                MongoDBSerializationHelper.fieldMapToBson(Collections.singletonMap(getIdField(), id))) > 0;
```

Comment:
You can probably had a protected helper method for `MongoDBSerializationHelper.fieldMapToBson(Collections.singletonMap(getIdField(), id)))` as it's used several times.